src/Coinbase/Exchange/Types/Socket.hs

```diff
@@ -122,225 +401,323 @@ data ExchangeMessage
+      "status" -> StatusMsg
+        <$> m .: "products"
+        <*> m .: "currencies"
+      "ticker" ->
+        let tickMsg = TickerMsg
+              <$> m .: "time"
+              <*> m .: "sequence"
+              <*> m .: "product_id"
+              <*> m .: "trade_id"
+              <*> m .: "price"
+              <*> m .: "side"
+              <*> m .: "last_size"
+              <*> m .:? "best_ask"
+              <*> m .:? "best_bid"
+            startTickerMsg = StartTickerMsg
+              <$> m .: "sequence"
+              <*> m .: "product_id"
+              <*> m .: "price"
+              <*> m .:? "best_ask"
+              <*> m .:? "best_bid"
+        in tickMsg <|> startTickerMsg
+      "snapshot" -> L2SnapshotMsg
+        <$> m .: "product_id"
+        <*> m .: "asks"
+        <*> m .: "bids"
+      "l2update" -> L2UpdateMsg
+        <$> m .: "time"
+        <*> m .: "product_id"
+        <*> m .: "changes"
+      "open" -> OpenMsg
+        <$> m .: "time"
+        <*> m .: "product_id"
+        <*> m .: "sequence"
+        <*> m .: "order_id"
+        <*> m .: "side"
+        <*> m .: "remaining_size"
+        <*> m .: "price"
+      "done" -> DoneMsg
+        <$> m .: "time"
+        <*> m .: "product_id"
+        <*> m .: "sequence"
+        <*> m .: "order_id"
+        <*> m .: "side"
+        <*> m .: "reason"
+        <*> m .:? "price"
+        <*> m .:? "remaining_size"
+      "match" -> MatchMsg  -- TODO: Four extra fields for authenticated
+                           -- user's matched orders; see docs
+        <$> m .: "time"
+        <*> m .: "product_id"
+        <*> m .: "sequence"
+        <*> m .: "side"
+        <*> m .: "trade_id"
+        <*> m .: "maker_order_id"
+        <*> m .: "taker_order_id"
+        <*> m .: "size"
+        <*> m .: "price"
+      "last_match" -> LastMatchMsg
+        <$> m .: "time"
+        <*> m .: "product_id"
+        <*> m .: "sequence"
+        <*> m .: "side"
+        <*> m .: "trade_id"
+        <*> m .: "maker_order_id"
+        <*> m .: "taker_order_id"
+        <*> m .: "size"
+        <*> m .: "price"
+      "change" -> do
+        ms <- m .:? "price"
+        let market = ChangeMarketMsg
+              <$> m .: "time"
+              <*> m .: "product_id"
+              <*> m .: "sequence"
+              <*> m .: "order_id"
+              <*> m .: "side"
+              <*> m .: "new_funds"
+              <*> m .: "old_funds"
+            limit = ChangeLimitMsg
+              <$> m .: "time"
+              <*> m .: "product_id"
+              <*> m .: "sequence"
+              <*> m .: "order_id"
+              <*> m .: "side"
+              <*> m .:? "price"
+              <*> m .: "new_size"
+              <*> m .: "old_size"
+        case (ms :: Maybe Price) of
+          Nothing -> market <|> limit
+          Just _ -> limit <|> market
+      "received" -> do
+        typ  <- m .:  "order_type"
+        mcid <- m .:? "client_oid" <|> pure Nothing
```

Comment:
Does this punch a hole that is too big? Specifically, does this accept messages that *do* have a `client_oid` that is not empty (nor the empty string), but that we can't parse?

Reply:
`m .:? "client_oid"` is `pure Nothing` if the "client_oid" field is missing, or `empty` if the field is present and the parse fails. So, yes, this results in `pure Nothing` exactly when the field is present and the parse fails.

I believe that's what we intended, see #4 and [my comment above](https://github.com/dimitri-xyz/haskell-coinbase-pro/pull/10#issuecomment-558950503). The problem was that we were getting runtime parsing failures upon receiving data with `client_oid` set to the empty string, which cannot be parsed into a UUID. Perhaps a better solution is to inspect the field, explicitly parse into `Nothing` when either the field is missing or set to the empty string, then parse as normal in the other case where the field is present but not an empty string? That would make true parsing failures result in an exception.

This is another thing we should notify Coinbase about, see #11. Presumably the `client_oid` field should be absent for orders that aren't the client's (and always absent when unauthenticated), but otherwise should always have an order ID. I wonder whether this is a bug in the server code in which the order ID field is created and populated because the order belongs to the client, but then it get's nulled (rather than removed) because the client isn't authenticated. ...

Reply:
It seems the hole is too big then. We want to allow only the situation when the `client_oid` field is *exactly* equal to an empty string `""` and nothing else, but this code (as before) also accepts when the `client_oid` is , for example, `"##$@%ˆ%%ˆ$E"`, which clearly should fail the parser and signal an error. I suggest we accept the code but also file/document the bug. Or fix it if possible!

I don't think Coinbase wants/should to clear the `client_oid` for unauthenticated users. It's extra (per connection) work to do the filtering. And the `client_oid` makes is easier to synchronize across multiple related accounts. So, customers should not be encouraged to rely on it and just do their own filtering.

Reply:
Looking at the docs, I think the explanation of the `client_oid` field may have been updated since last I read it:

> CLIENT ORDER ID
> 
> The optional client_oid field must be a UUID generated by your trading application. This field value will be broadcast in the public feed for received messages. You can use this field to identify your orders in the public feed.
> 
> The client_oid is different than the server-assigned order id. If you are consuming the public feed and see a received message with your client_oid, you should record the server-assigned order_id as it will be used for future order status updates. The client_oid will NOT be used after the received message is sent.
> 
> The server-assigned order id is also returned as the id field to this HTTP POST request.

So `client_oid` should be present for all orders whose submitters supplied a (valid?) UUID, not just for the client's own orders. This doesn't really help us understand why the field is set to the empty string sometimes. Perhaps that's what Coinbase sets the field to when a client assigns an invalid UUID? In any event, I think it makes sense to re-write this to parse as we've discussed.